torch/utils/data/datapipes/iter/combining.py

```diff
@@ -362,6 +368,7 @@ def _find_next(self, instance_id: int) -> T_co:
     def get_next_element_by_instance(self, instance_id: int):
         if self._datapipe_iterator is None and not self.main_datapipe_exhausted:
             self._datapipe_iterator = iter(self.main_datapipe)
+            self._snapshot_state = _SnapshotState.Iterating  # This is necessary for the DataPipe to reset properly.
```

Comment:
Yes, because creating a new iterator is always allowed but using an invalidated iterator isn not allowed. I believe `self._snapshot_state` can tell us whether or not an iterator is created, but not which iterator is valid.

Reply:
Oh yea. Forget datapipe doesn't keep track of its iterators. NVM